### ISSUE TITLE
Remove `onehotbatch`?

### ISSUE THREAD

**Issue by mcabbott** [FluxML/Flux.jl#1659]
Since I forgot this and just stumbled over it. Why aren't `onehot` and `onehotbatch` the same function? It seems like there's a completely natural function that always increases `ndims` by 1, but somehow it has two names:
```julia
julia> Flux.onehot('a', 'a':'z') |> summary
"26-element Flux.OneHotArray{UInt32, 26, 0, 1, UInt32}"

julia> Flux.onehotbatch(collect("hello"), 'a':'z') |> summary
"26×5 Flux.OneHotArray{UInt32, 26, 1, 2, Vector{UInt32}}"

julia> Flux.onehotbatch(['a' 'b'; 'c' 'd'], 'a':'z') |> summary
"26×2×2 Flux.OneHotArray{UInt32, 26, 2, 3, Matrix{UInt32}}"
```

**Comment by darsnack** [comment]
Yeah appears to be a legacy function at this point.

**Comment by mcabbott** [comment]
Not certain this is what you're saying, but if someone has a set of labels which are themselves (say) vectors, then `onehot(vector , labels)` will now return one one-hot vector. That should probably still work. Can we rely on `eltype(labels)` to tell us?

What would have to change is things like the first example below, which I think we can safely declare to be a mistake (works only as default has a different type):
```julia
julia> Flux.onehot(collect("hello world"), 'a':'z', 'a')
26-element Flux.OneHotVector{26,UInt32}:
 1
 0
 0
 0
...

julia> Flux.onehotbatch(collect("hello world"), 'a':'z', 'a')
26×11 Flux.OneHotArray{26,2,Vector{UInt32}}:
 0  0  0  0  0  1  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  1
 0  1  0  0  0  0  0  0  0  0  0
...
```

**Comment by CarloLucibello** [comment]
In #223 Mike mentioned 2 corner cases
```
What if you are single encoding a vector? What if you are batch encoding an iterator / non-vector collection?
```

**Comment by darsnack** [comment]
Yeah the first corner case is what I was trying to get at. But after sleeping on it, I'm not sure it is an edge case?

**Comment by mcabbott** [comment]
Ok, thanks. My thinking is that we can work around these edge-cases, by checking eltypes / the type of the first element. Although I haven't tried to write it. 

"single encoding a vector" means the labels has to be a container of vectors, which should be easy to see.

"batch encoding an iterator" does work, right now. To keep this working for arbitrary custom iterators, one path has to just call `iterate` on the 1st argument. Maybe the hard edge case is `onehot(x::Weird, labs::Vector{OtherWeird})` where `Weird` does not support iteration, but does support `isequal(::Weird, :: OtherWeird). 
```
julia> onehotbatch((mod1(i,3) for i in 1:10 if i < 8), 1:3) # generator of batch elements
3×7 OneHotMatrix(::Vector{UInt32}) with eltype Bool:
 1  ⋅  ⋅  1  ⋅  ⋅  1
 ⋅  1  ⋅  ⋅  1  ⋅  ⋅
 ⋅  ⋅  1  ⋅  ⋅  1  ⋅

julia> onehot([2,4], [i,i^2] for i in 1:4)  # simple generator of labels
4-element OneHotVector(::UInt32) with eltype Bool:
 ⋅
 1
 ⋅
 ⋅

julia> onehot([2,4], [i,i^2] for i in 1:4 if i^2 < 15)  # generator of labels -- doesn't support `findfirst`
ERROR: MethodError: no method matching keys(::Base.Iterators.Filter{var"#8#10", UnitRange{Int64}})
```

**Comment by CarloLucibello** [comment]
Closing — `onehot`/`onehotbatch` moved to OneHotArrays.jl, so this is no longer Flux`s call, and no consensus was reached here on resolving the ambiguous cases (encoding a single vector-valued label vs. batch-encoding a collection). Removing or merging the two would be a breaking change to a widely used API for a mostly aesthetic gain.
